### PR TITLE
Update pruning IR pass.

### PR DESCRIPTION
**Pull request by fodinabor**
This cleans up the function pruning and moves the globals pruning into an accordingly named pass.
To support the new pass manager infrastructure which will be the default for Clang 13+.
The new PM can be activated in Clang 11/12 with -fexperimental-new-pass-manager.
The option -fpass-plugin= is already present in Clang 11.

### CODE REVIEW

**Comment by illuhad** [comment]
@hipSYCL-ci-bot test

File: src/compiler/HipsyclClangPlugin.cpp [hunk 1]
```diff
@@ -24,35 +24,49 @@
...
+static llvm::RegisterStandardPasses RegisterGlobalsPruningPassOptimizerLast(llvm::PassManagerBuilder::EP_OptimizerLast,
+                                                                            registerGlobalsPruningPass);
 
-static llvm::RegisterStandardPasses
-  RegisterFunctionPruningIRPassOptimizerLast(llvm::PassManagerBuilder::EP_OptimizerLast,
-                                             registerFunctionPruningIRPass);
+#ifndef _WIN32
+#define HIPSYCL_STRINGIFY(V) #V
+#define HIPSYCL_PLUGIN_VERSION_STRING                                                                                  \
+  "v" HIPSYCL_STRINGIFY(HIPSYCL_VERSION_MAJOR) "." HIPSYCL_STRINGIFY(HIPSYCL_VERSION_MINOR)
```
Comment (illuhad) [inline src/compiler/HipsyclClangPlugin.cpp]: What about the patch version level?

File: src/compiler/HipsyclClangPlugin.cpp [hunk 2]
```diff
@@ -24,35 +24,49 @@
...
+
+extern "C" LLVM_ATTRIBUTE_WEAK ::llvm::PassPluginLibraryInfo llvmGetPassPluginInfo() {
+  return {LLVM_PLUGIN_API_VERSION, "hipSYCL Clang plugin", HIPSYCL_PLUGIN_VERSION_STRING, [](llvm::PassBuilder &PB) {
+            // Note: for Clang < 12, this EP is not called for O0, but the new PM isn't really used there anyways..
+            PB.registerOptimizerLastEPCallback([](llvm::ModulePassManager &MPM, llvm::PassBuilder::OptimizationLevel) {
+              MPM.addPass(hipsycl::compiler::GlobalsPruningPass{});
```
Comment (illuhad) [inline src/compiler/HipsyclClangPlugin.cpp]: Please restrict line length.. I think most of the code uses something like 80 or 90 characters

**Review by illuhad — COMMENTED** [review]
So does this break compatibility with clang pre-11?

**Comment by fodinabor** [comment]
Hmm... just tested with Clang 11-13, just looked up `fpass-plugin` and it seems to be present from Clang 9 onwards.
I would have to check whether the new PM code works with Clang 9. From a quick code search, it seems like the infrastructure used here should've been there already -> should compile. Don't really fancy getting another LLVM version locally to test, though... if it goes wrong, we could always extend the `_WIN32` macros there to check the LLVM version as well.

**Comment by fodinabor** [comment]
@hipsycl-ci-bot test after rebase.

**Review by illuhad — APPROVED** [review]
@fodinabor okay, feel free to merge at your discretion :)

**Comment by psalz** [comment]
> 
> 
> So does this break compatibility with clang pre-11?

It would seem so. Clang 10:

```
src/compiler/HipsyclClangPlugin.cpp:71:17: error: no viable conversion from '(lambda at src/compiler/HipsyclClangPlugin.cpp:71:17)' to 'const std::function<void (FunctionPassManager &, OptimizationLevel)>' (aka 'const function<void (PassManager<llvm::Function> &, llvm::PassBuilder::OptimizationLevel)>')
                [](llvm::ModulePassManager &MPM, llvm::PassBuilder::OptimizationLevel) {
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/std_function.h:403:7: note: candidate constructor not viable: no known conversion from '(lambda at src/compiler/HipsyclClangPlugin.cpp:71:17)' to 'std::nullptr_t' (aka 'nullptr_t') for 1st argument
      function(nullptr_t) noexcept
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/std_function.h:414:7: note: candidate constructor not viable: no known conversion from '(lambda at src/compiler/HipsyclClangPlugin.cpp:71:17)' to 'const std::function<void (llvm::PassManager<llvm::Function> &, llvm::PassBuilder::OptimizationLevel)> &' for 1st argument
      function(const function& __x);
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/std_function.h:423:7: note: candidate constructor not viable: no known conversion from '(lambda at src/compiler/HipsyclClangPlugin.cpp:71:17)' to 'std::function<void (llvm::PassManager<llvm::Function> &, llvm::PassBuilder::OptimizationLevel)> &&' for 1st argument
      function(function&& __x) noexcept : _Function_base()
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/std_function.h:447:2: note: candidate template ignored: requirement '_Callable<(lambda at src/compiler/HipsyclClangPlugin.cpp:71:17), std::__invoke_result<(lambda at src/compiler/HipsyclClangPlugin.cpp:71:17) &, llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>> &, llvm::PassBuilder::OptimizationLevel> >::value' was not satisfied [with _Functor = (lambda at src/compiler/HipsyclClangPlugin.cpp:71:17), $1 = void]
        function(_Functor);
        ^
src/compiler/HipsyclClangPlugin.cpp:71:17: note: candidate function
                [](llvm::ModulePassManager &MPM, llvm::PassBuilder::OptimizationLevel) {
                ^
/usr/lib/llvm-10/include/llvm/Passes/PassBuilder.h:559:76: note: passing argument to parameter 'C' here
      const std::function<void(FunctionPassManager &, OptimizationLevel)> &C)
```

**Pull request merged**